### ISSUE TITLE
Provide also traditional sessions?

### ISSUE THREAD

**Issue by Ocramius** [Ocramius/PSR7Session#21]
Since we renamed the repo from `StorageLessSession` to `PSR7Session`, we may as well introduce support for a "standard" session middleware (with session IDs and server-side session storage):

 - UUID based (no low crypto ID generation)
 - file-only support first (like `ext/session`)
 - same concepts as `ext/session` (validity of entries, etc)
 - clear adapter specification (stuff like GC should be a detail)
 - API to force session ID regeneration
 - same session container stuff that we currently have

**Comment by Ocramius** [comment]
Possible storage adapter interface (requires further inspecting):

```php
interface SessionStorage
{
    public function save(SessionInterface $session);
    public function load(SessionIdInterface $id) : SessionInterface;
    public function destroy(SessionIdInterface $id);
}
```

I would still apply the rules around the contents (do not allow object serialization).

**Comment by danizord** [comment]
:+1:

**Comment by shochdoerfer** [comment]
:+1:

**Comment by MidnightDesign** [comment]
Sounds great. Would I be able to use this (easily) an an MVC context? Some class that knows about the storage and pulls the session data from there or creates a new session if no session cookie is set.

**Comment by Ocramius** [comment]
>  Some class that knows about the storage and pulls the session data from there or creates a new session if no session cookie is set.

Probably not - the problem with this kind of approach is that the session system would need to be aware of the "current" request.

Something like you proposed in #zendframework/zend-session#11 could work:

```php
$sessionContainer = $sessionManager->fromRequest($request);
```

**Comment by MidnightDesign** [comment]
The only problem with my proposal is how to get the newly created session ID into the response. That would have to be done manually after obtaining the new session container, wouldn't it?

**Comment by Ocramius** [comment]
@MidnightDesign I suggest looking at the current middleware for an idea of how this is achieved: the middleware wraps the entire app and can decide when a session needs to be written to disk.

**Comment by MidnightDesign** [comment]
Yeah, I get how this works. And I love Middleware. But I'm thinking about other contexts right now - in my case ZF2. That wouldn't be possible there, would it? Well, maybe with events...

When do you think will you be able to start working on that stuff? I could try implementing a ZF2 module based on your work. Can I help you with anything on this project?

**Comment by Ocramius** [comment]
@MidnightDesign if you could try implementing the storage-based sessions, that would be rad.

It shouldn't be too complex

**Comment by MidnightDesign** [comment]
I've never contributed any larger chunks of code to an open source project before (just little bugfixes and stuff), but I'm be happy to give it a shot.

**Comment by Ocramius** [comment]
@MidnightDesign this issue is marked for 2.0.0 - you are free to do whatever you want on it :-)

Suggestion: open a PR early, so that it can be discussed carefully.

**Comment by MidnightDesign** [comment]
I'll just open the PR with only the interfaces you mentioned we'll see where it goes from there. Should I open it against master or are you going to create a 2.0 branch?

**Comment by Ocramius** [comment]
@MidnightDesign `master`, thanks! :-)

**Comment by liuggio** [comment]
just a thought
in other languages "session" even don't exists
and be a no-state application is a great goal for new app (and this lib is made for PHP7 app)...
session with no storage is a great way to push the good practice of not using the session at all,
and use the psr interface only for auth. 

I'm not saying you shouldn't are just few thoughts.

**Comment by Ocramius** [comment]
Closing as per https://github.com/psr7-sessions/storageless/pull/30#issuecomment-940070995
